File: serverless_sim_ui/src/apis.ts
import axios from "axios"



export class GetNetworkTopoRespExist {
    constructor(
        public topo:number[][],
    ){}
}

export class GetNetworkTopoRespNotFound {
    constructor(
        public msg:string,
    ){}
}

export class GetNetworkTopoResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    exist():undefined| GetNetworkTopoRespExist{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
    not_found():undefined| GetNetworkTopoRespNotFound{
        if(this.id==2){
            return this.kernel
        }
        return undefined
    }
    
}


export class GetNetworkTopoReq {
    constructor(
        public env_id:string,
    ){}
}

export namespace apis {
    export async function get_network_topo(req:GetNetworkTopoReq):Promise<GetNetworkTopoResp>{
        let res:any = await axios.post("/api/get_network_topo", req)
        return new GetEnvIdResp(res.data.kernel,res.data.id)
    }
}




export class GetEnvIdRespExist {
    constructor(
        public env_id:string[],
    ){}
}

export class GetEnvIdRespNotFound {
    constructor(
        public msg:string,
    ){}
}

export class GetEnvIdResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    exist():undefined| GetEnvIdRespExist{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
    not_found():undefined| GetEnvIdRespNotFound{
        if(this.id==2){
            return this.kernel
        }
        return undefined
    }
    
}


export namespace apis {
    export async function get_env_id():Promise<GetEnvIdResp>{
        let res:any = await axios.post("/api/get_env_id", )
        return new GetEnvIdResp(res.data.kernel,res.data.id)
    }
}




export class ResetRespSuccess {
    constructor(
        public env_id:string,
    ){}
}

export class ResetRespInvalidConfig {
    constructor(
        public msg:string,
    ){}
}

export class ResetResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    success():undefined| ResetRespSuccess{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
    invalid_config():undefined| ResetRespInvalidConfig{
        if(this.id==2){
            return this.kernel
        }
        return undefined
    }
    
}


export class ResetReq {
    constructor(
        public config:any,
    ){}
}

export namespace apis {
    export async function reset(req:ResetReq):Promise<ResetResp>{
        let res:any = await axios.post("/api/reset", req)
        return new GetEnvIdResp(res.data.kernel,res.data.id)
    }
}




export class StepRespSuccess {
    constructor(
        public state:string,
        public score:number,
        public stop:boolean,
        public info:string,
    ){}
}

export class StepRespEnvNotFound {
    constructor(
        public msg:string,
    ){}
}

export class StepResp{
    constructor(
        private kernel: any,
        private id: number
    ) {}
    
    success():undefined| StepRespSuccess{
        if(this.id==1){
            return this.kernel
        }
        return undefined
    }
    
    env_not_found():undefined| StepRespEnvNotFound{
        if(this.id==2){
            return this.kernel
        }
        return undefined
    }
    
}


export class StepReq {
    constructor(
        public env_id:string,
        public action:number,
    ){}
}

export namespace apis {
    export async function step(req:StepReq):Promise<StepResp>{
        let res:any = await axios.post("/api/step", req)
        return new GetEnvIdResp(res.data.kernel,res.data.id)
    }
}

File: serverless_sim/src/apis.rs

use serde_json::{json,Value};
use serde::{Serialize, Deserialize};
use axum::{http::StatusCode, routing::post, Json, Router};
use async_trait::async_trait;
use crate::network::ApiHandlerImpl;


#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum GetNetworkTopoResp{
    Exist{
       topo:Vec<Vec<f64>>,
    },

    NotFound{
       msg:String,
    },

}

impl GetNetworkTopoResp {
    fn id(&self)->u32 {
        match self {
                GetNetworkTopoResp::Exist{..}=>1,
                GetNetworkTopoResp::NotFound{..}=>2,
            }
    }

    pub fn serialize(&self)->Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}


#[derive(Debug, Serialize, Deserialize)]
pub struct GetNetworkTopoReq {
    pub env_id:String,
}



#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum GetEnvIdResp{
    Exist{
       env_id:Vec<String>,
},
    NotFound{
       msg:String,
},

}

impl GetEnvIdResp {
    fn id(&self)->u32 {
        match self {
                GetEnvIdResp::Exist{..}=>1,
    GetEnvIdResp::NotFound{..}=>2,

        }
    }
    pub fn serialize(&self)->Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}




#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ResetResp{
    Success{
       env_id:String,
},
    InvalidConfig{
       msg:String,
},

}

impl ResetResp {
    fn id(&self)->u32 {
        match self {
                ResetResp::Success{..}=>1,
    ResetResp::InvalidConfig{..}=>2,

        }
    }
    pub fn serialize(&self)->Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}


#[derive(Debug, Serialize, Deserialize)]
pub struct ResetReq {
       pub config:Value,
}



#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum StepResp{
    Success{
       state:String,
       score:f64,
       stop:bool,
       info:String,
},
    EnvNotFound{
       msg:String,
},

}

impl StepResp {
    fn id(&self)->u32 {
        match self {
                StepResp::Success{..}=>1,
    StepResp::EnvNotFound{..}=>2,

        }
    }
    pub fn serialize(&self)->Value {
        json!({
            "id": self.id(),
            "kernel": serde_json::to_value(self).unwrap(),
        })
    }
}


#[derive(Debug, Serialize, Deserialize)]
pub struct StepReq {
    pub env_id:String,
    pub action:i32,
}


#[async_trait]
pub trait ApiHandler {
    
    async fn handle_get_network_topo(&self, req:GetNetworkTopoReq)->GetNetworkTopoResp;
            
    async fn handle_get_env_id(&self, )->GetEnvIdResp;
            
    async fn handle_reset(&self, req:ResetReq)->ResetResp;
            
    async fn handle_step(&self, req:StepReq)->StepResp;
            
}


pub fn add_routers(mut router:Router)->Router
{
    
    async fn get_network_topo(Json(req):Json<GetNetworkTopoReq>)-> (StatusCode, Json<Value>){
        (StatusCode::OK, Json(ApiHandlerImpl.handle_get_network_topo(req).await.serialize()))
    }
    router=router
        .route("/get_network_topo", post(get_network_topo));
                             
    async fn get_env_id()-> (StatusCode, Json<Value>){
        (StatusCode::OK, Json(ApiHandlerImpl.handle_get_env_id().await.serialize()))
    }
    router=router
        .route("/get_env_id", post(get_env_id));

    // python脚本文件向前端请求的接口，前端调用的 reset 函数，用于创建模拟环境对象
    async fn reset(Json(req):Json<ResetReq>)-> (StatusCode, Json<Value>){
        (StatusCode::OK, Json(ApiHandlerImpl.handle_reset(req).await.serialize()))
    }
    router=router
        .route("/reset", post(reset));

    // python脚本文件向前端请求的接口，前端调用的 step 函数   
    async fn step(Json(req):Json<StepReq>)-> (StatusCode, Json<Value>){
        (StatusCode::OK, Json(ApiHandlerImpl.handle_step(req).await.serialize()))
    }
    router=router
        .route("/step", post(step));
                             
    
    router
}

